### PR TITLE
Convert TestUtils.java into kotlin

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Fixes: #3050 
Parent Issue: #1826 

Converted TestUtils.java into kotlin

### CODE REVIEW

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  @JvmStatic fun withContent(content: String): Matcher<Any?> {
+    return object : BoundedMatcher<Any?, Any?>(Any::class.java) {
+      public override fun matchesSafely(myObj: Any?): Boolean {
+        if (myObj !is LibraryNetworkEntity.Book) {
+          return false
+        }
+        return if (myObj.getUrl() != null) {
+          myObj.getUrl().contains(content)
+        } else {
+          myObj.file.path.contains(content)
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:124]: We deprecated the `File` object in `LibraryNetworkEntitiy`. Perhaps it's time to remove deprecated objects/methods.
https://github.com/kiwix/kiwix-android/blob/2ab7a552d7b763b342f64a88956607ad0a0e85e6/core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java#L94-L97

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    val outFile = File(screenshotDir.path + File.separator + fileName)
+    val screenshot = Screenshot.capture().bitmap ?: return
+    try {
+      val fos = FileOutputStream(outFile)
+      screenshot.compress(Bitmap.CompressFormat.PNG, 90, fos)
+      fos.close()
+    } catch (e: FileNotFoundException) {
+      Log.w(TAG, "Failed to save Screenshot", e)
+    } catch (e: IOException) {
+      Log.w(TAG, "Failed to save Screenshot", e)
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:111]: Perhaps a more meaningful message in each exception

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      }
+    }
+    val timestamp = SimpleDateFormat("ddMMyyyy_HHmm").format(Date())
+    val fileName = "TEST_${timestamp}_$name.png"
+    val outFile = File(screenshotDir.path + File.separator + fileName)
+    val screenshot = Screenshot.capture().bitmap ?: return
+    try {
+      val fos = FileOutputStream(outFile)
+      screenshot.compress(Bitmap.CompressFormat.PNG, 90, fos)
+      fos.close()
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:107]: We should close in the `finally` block.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  @JvmStatic fun allowPermissionsIfNeeded() {
+    if (Build.VERSION.SDK_INT >= 23 && !hasStoragePermission()) {
+      val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+      val allowPermissions =
+        device.findObject(
+          UiSelector().clickable(true)
+            .checkable(false).index(1)
+        )
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:80]: This function's name is one thing and does the other thing. We can remove this generic functionality and split it into small generic functions. The naming of functionality will mislead us in the future.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        however I'm unsure if
+    it's speed related, or Android Version related.
+   */
+  private fun hasStoragePermission(): Boolean {
+    return ContextCompat.checkSelfPermission(
+      InstrumentationRegistry.getInstrumentation().targetContext,
+      Manifest.permission.READ_EXTERNAL_STORAGE
+    ) == PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(
+      InstrumentationRegistry.getInstrumentation().targetContext,
+      Manifest.permission.WRITE_EXTERNAL_STORAGE
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:69]: Here too, we can add three more small helper functions a. read external storage, b. write external storage, c. manage external storage. Also, we can put this into another helper function that checks all of them.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    The number 250 is fairly arbitrary. I found 100 to be insufficient,
+        and 250 seems to work on all devices I've tried.
+
+    The sleep combats an intermittent issue caused by
+        tests executing before the app/activity is ready.
+    This isn't necessary on all devices (particularly more recent ones),
+        however I'm unsure if
+    it's speed related, or Android Version related.
+   */
+  private fun hasStoragePermission(): Boolean {
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:63]: We should scope this functions to `public` or `internal`

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+      val allowPermissions =
+        device.findObject(
+          UiSelector().clickable(true)
+            .checkable(false).index(1)
+        )
+      if (allowPermissions.exists()) {
+        try {
+          allowPermissions.click()
+        } catch (e: UiObjectNotFoundException) {
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:84]: We should log this important exception as well.

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 6 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 5 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 4 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 3 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 2 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: hi @gouri-panda 
done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 1 above)
Comment (gouri-panda) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:135]: We should find an alternative for this, not removing will help us :). Maybe it's worth a new ticket.

Comment (gouri-panda) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:135]: #3092 will fix this

File: core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java [hunk 8]
```diff
@@ -91,7 +91,7 @@ public static class Book implements Serializable {
     public String tags;
 
     public int searchMatches = 0;
-    @Deprecated
+
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java:94]: Revert this we will fix this in #3092

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  private fun hasManageExternalStoragePermission(): Boolean =
+    Environment.isExternalStorageManager()
+
+  @JvmStatic fun hasStoragePermission(): Boolean =
+    if (Build.VERSION.SDK_INT < 23) true
+    else hasReadExternalStoragePermission() && hasWriteExternalStoragePermission()
+
+  @JvmStatic fun allowStoragePermission() {
+    val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+    val allowPermissions =
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:87]: We should check here if permission has already been granted or not. Or we can hoist this to another functionality as we did before `allowPermissionsIfNeeded()`.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      Manifest.permission.WRITE_EXTERNAL_STORAGE
+    ) == PackageManager.PERMISSION_GRANTED
+
+  @RequiresApi(Build.VERSION_CODES.R)
+  private fun hasManageExternalStoragePermission(): Boolean =
+    Environment.isExternalStorageManager()
+
+  @JvmStatic fun hasStoragePermission(): Boolean =
+    if (Build.VERSION.SDK_INT < 23) true
+    else hasReadExternalStoragePermission() && hasWriteExternalStoragePermission()
```
Comment (gouri-panda) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt:83]: [nit] We can inline this function and get rid of this if else statement
```
@JvmStatic fun hasStoragePermission() = Build.VERSION.SDK_INT > 23 && hasReadExternalStoragePermission() && hasWriteExternalStoragePermission() 
```

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt (hunk 9 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt]: done

File: core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java (hunk 8 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java]: done

**Review by gouri-panda — APPROVED** [review]
Thanks @MohitMaliDeveloper !

**Pull request merged**